### PR TITLE
Streaming audio files

### PR DESCRIPTION
**Pull request by ShFil119**
To think about:

- [x] smarter use of mutexes
- [x] testing with thread sanitizer
- [x] OpenAL issue:
 ```
OpenAL error at ../rwengine/src/audio/SoundBufferStreamed.cpp:22: Invalid value!
OpenAL error at ../rwengine/src/audio/SoundBufferStreamed.cpp:23: Invalid operation!
```

### CODE REVIEW

File: rwengine/src/audio/SoundBufferStreamed.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    void play() final;
+    void pause() final;
+    void stop() final;
+
+private:
+    SoundSource* soundSource = nullptr;
+    void updateBuffers();
+    unsigned int streamedData = 0;
+    std::array<ALuint, kNrBuffersStreaming> buffers;
+    std::future<void> loadingThread;
```
Comment (danhedron) [inline rwengine/src/audio/SoundBufferStreamed.hpp]: Nit naming wise: `SoundSource` is doing the loading, the async code here just uploads the data into OpenAL right?

File: rwengine/src/audio/SoundBufferStreamed.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    loadingThread = std::async(std::launch::async,
+                               &SoundBufferStreamed::updateBuffers, this);
+}
+
+void SoundBufferStreamed::updateBuffers() {
+    while (true) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+        std::lock_guard<std::mutex> lock(soundSource->mutex);
+        if (!running) {
```
[the remark this answers is not in the document]
Comment (ShFil119) [reply rwengine/src/audio/SoundBufferStreamed.cpp]: Firstly `~Sound` should call stop on buffer // see above
And after calls `~SoundBufferStreamed`. // Maybe 100 ms delay will happen.

File: rwengine/src/audio/SoundBufferStreamed.hpp (hunk 1 above)
Comment (ShFil119) [reply rwengine/src/audio/SoundBufferStreamed.hpp]: It feeds buffers and queue them (to be used).

Good point. :)

File: rwengine/src/audio/SoundBuffer.hpp [hunk 3]
```diff
@@ -4,22 +4,24 @@
 #include <al.h>
 #include <glm/vec3.hpp>
 
+#include <vector>
```
[the remark this answers is not in the document]
Comment (ShFil119) [reply rwengine/src/audio/SoundBuffer.hpp]: yeah, to remove

File: rwengine/src/audio/SoundBufferStreamed.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (danhedron) [reply rwengine/src/audio/SoundBufferStreamed.cpp]: [backtrace](https://gist.github.com/danhedron/36cc8414973c72472c85cd21910476d6).

Due to another bug the opening mission fails as soon as the kuruma spawns, and the game then hangs with the backtrace above.

[the remark this answers is not in the document]
Comment (danhedron) [reply rwengine/src/audio/SoundBufferStreamed.cpp]: Narrowed it down, [this diff](https://gist.github.com/danhedron/620108474813175179d20e7a7258417b) will hang the data tests.

[the remark this answers is not in the document]
Comment (ShFil119) [reply rwengine/src/audio/SoundBufferStreamed.cpp]: Only this test hangs? Or rwgame also?

**Pull request merged**